### DISCUSSION TITLE
Possible to remove ajv and pino?

### DISCUSSION THREAD

**Discussion by prescience-data** [fastify/fastify#4339] [Q&A]
In all other parts of the application, we use `Zod` and `tslog` which would be trivial to provide as interfaces that mimic `avj` and `pino`. 
My understanding is this is possible for `pino`,  however, I've never been able to treeshake it out fully. (source: https://www.fastify.io/docs/latest/Reference/Logging/)

Looking at the total bundle size, `avj` and `pino` add a huge amount of weight (>15% unmodified, >16% minified) so it would be really nice to be able to totally omit them from the bundle and provide the existing `Zod` and `tslog` libraries in their place. 

![image](https://user-images.githubusercontent.com/65471523/195973588-ede3f533-be03-4602-ada5-1d177c967699.png)

Even when logging is set to false, the `pino` package still won't treeshake out, which is strange given the wording of the docs around "abstract logging". I feel like I might be doing something wrong here, so would love to hear any methods others have used to potentially achieve this!

_(PS: This isn't a value/quality question about any of the libraries mentioned, just that it's unfortunate the bundle needs to blow out with redundant/duplicated concerns.)_

**Comment by mcollina** [comment]
My recommendation is to not worry about bundle size in Node.js server-side application. It mostly does not matter.

(There are a few exceptions where this matters, however you are better off with Ajv and pino in those cases).

Ajv is used internally. I don't think is possible to swap it out. Pino might be removed with some webpack-config.

**Comment by Eomm** [comment]
It is doable and I found a bug in Fastify 🤣

How to do it:

1) use the package.json's `overrides` property:

```
{
  "name": "ciao",
  "version": "1.0.0",
  "main": "index.js",
  "overrides": {
    "@fastify/ajv-compiler": "npm:empty@0.10.1",
    "@fastify/fast-json-stringify-compiler": "npm:empty@0.10.1"
  },
  "dependencies": {
    "fastify": "^4.8.1"
  }
}
```

Note that I set `npm:empty`. This is a real npm module, you can define or publish an empty module just to download its payload or link a github repository.

After `npm install` you will see:

```
➜  ciao > npm ls ajv 
ciao@1.0.0 /Users/mspigolon/workspace/_experiments/ciao
└── (empty)
```

2) take ownership of the `schemaController`:

```
const fastify = require('fastify')({ 
  logger: true,
  schemaController: {
    compilersFactory: {
      buildValidator: function factory (externalSchemas, ajvServerOption) {
        return function validatorCompiler ({ schema, method, url, httpPart }) {
          return () => {return true}
        }
      },
      buildSerializer: function factory (externalSchemas, serializerOptsServerOption) {
        return function serializerCompiler ({ schema, method, url, httpStatus }) {
          return () => { return true}
        }
      }
    }
  }
})

fastify.get('/', async (request, reply) => {
  return { hello: 'world' }
})

fastify.listen(8080)
```

If you start it, it will work after the fix (because right now the default `@fastify/ajv-compiler` is called even if you customize it.

I tried locally the fix and it works

**Reply by mcollina** [reply]
I think Ajv is still used by fast-json-stringify and light-my-request.

**Reply by Eomm** [reply]
Looking at the dep tree, we are free from ajv by removing the 2 deps I wrote above.

I removed the ajv dep from light-my-request some time ago: https://github.com/fastify/light-my-request/pull/195


Meanwhile, here is the fix: https://github.com/fastify/fastify/pull/4340

**Reply by mcollina** [reply]
Ah I forgot about that.
